### PR TITLE
Shim convert for transform subcommand

### LINKED ISSUE #26: Add transform windup subcommand to convert XML rules to YAML

### PR DESCRIPTION
**Pull request by eemcmullan**
closes #26

### CODE REVIEW

File: cmd/shimconvert.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	return windupShimCommand
+}
+
+func (w *windupShimCommand) Validate() error {
+	outputStat, err := os.Stat(w.output)
+	if err != nil {
+		return err
+	}
+	if !outputStat.IsDir() {
+		log.Errorf("output path %s is not a directory", w.output)
```
Comment (pranavgaikwad) [inline cmd/shimconvert.go:55]: use `w.log`

File: cmd/shimconvert.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	}
+	if w.input == nil || len(w.input) == 0 {
+		return fmt.Errorf("input for rule file or directory must not be empty")
+	}
+	return nil
+}
+
+func (w *windupShimCommand) getRulesVolumes(tempRuleDir string) (map[string]string, error) {
+	rulesVolumes := make(map[string]string)
+	err := os.Mkdir(tempRuleDir, os.ModePerm)
```
Comment (pranavgaikwad) [inline cmd/shimconvert.go:66]: can we use `MkdirTemp` like https://github.com/konveyor-ecosystem/kantra/blob/3f97febd62adfbdcc57ebbe9ae583a474c08d51c/cmd/analyze.go#L363? and then clean up all dirs in PostRunE

File: cmd/shimconvert.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		} else {
+			tempRuleDir = r
+		}
+	}
+	rulesVolumes[tempRuleDir] = XMLRulePath
+	return rulesVolumes, nil
+}
+
+func (w *windupShimCommand) Run(ctx context.Context) error {
+	wd, err := os.Getwd()
```
Comment (pranavgaikwad) [inline cmd/shimconvert.go:94]: no need to do this, if we use `MkdirTemp`

File: cmd/shimconvert.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+func (w *windupShimCommand) getRulesVolumes(tempRuleDir string) (map[string]string, error) {
+	rulesVolumes := make(map[string]string)
+	err := os.Mkdir(tempRuleDir, os.ModePerm)
+	if err != nil {
+		return nil, err
+	}
+	for _, r := range w.input {
+		stat, err := os.Stat(r)
+		if err != nil {
+			log.Errorf("failed to stat rules %s", r)
```
Comment (pranavgaikwad) [inline cmd/shimconvert.go:73]: use `w.log`

File: cmd/shimconvert.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+			return nil, err
+		}
+		// move xml rule files from user into dir to mount
+		if !stat.IsDir() {
+			xmlFileName := filepath.Base(r)
+			destFile := filepath.Join(tempRuleDir, xmlFileName)
+			err := copyFileContents(r, destFile)
+			if err != nil {
+				log.Errorf("failed to move rules file from %s to %s", r, destFile)
```
Comment (pranavgaikwad) [inline cmd/shimconvert.go:82]: use `w.log`

File: cmd/shimconvert.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+				return nil, err
+			}
+		} else {
+			tempRuleDir = r
+		}
+	}
+	rulesVolumes[tempRuleDir] = XMLRulePath
```
Comment (pranavgaikwad) [inline cmd/shimconvert.go:89]: I may be wrong but this doesn't look right, when there are multiple --input options, it looks like we will always just return the last one set in the loop?

**Review by pranavgaikwad — CHANGES_REQUESTED** [review]

File: cmd/shimconvert.go (hunk 2 above)
Comment (pranavgaikwad) [reply cmd/shimconvert.go:66]: please ignore my previous comment, I think it's easier to just run it in workdir

File: cmd/shimconvert.go (hunk 3 above)
Comment (pranavgaikwad) [reply cmd/shimconvert.go:94]: please ignore, I see that this makes it easier to cleanup later ahead in the same function

File: cmd/shimconvert.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	log logr.Logger
+}
+
+func NewWindupShimCommand(log logr.Logger) *cobra.Command {
+	windupShimCmd := &windupShimCommand{
+		log: log,
+	}
+
+	windupShimCommand := &cobra.Command{
+		Use: "xmlrules",
```
Comment (pranavgaikwad) [inline cmd/shimconvert.go:28]:
```suggestion
		Use: "rules",
```

File: cmd/shimconvert.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+			err = windupShimCmd.Run(cmd.Context())
+			if err != nil {
+				return err
+			}
+			return nil
+		},
+	}
+	windupShimCommand.Flags().StringArrayVarP(&windupShimCmd.input, "input", "i", []string{}, "path to XML rule file(s) or directory")
+	windupShimCommand.Flags().StringVarP(&windupShimCmd.output, "output", "o", "", "path to output directory")
+
```
Comment (pranavgaikwad) [inline cmd/shimconvert.go:45]: you can mark these flags required like https://github.com/konveyor-ecosystem/kantra/blob/3f97febd62adfbdcc57ebbe9ae583a474c08d51c/cmd/analyze.go#L75

**Review by pranavgaikwad — CHANGES_REQUESTED** [review]

File: cmd/shimconvert.go (hunk 6 above)
Comment (eemcmullan) [reply cmd/shimconvert.go:89]: In the loop, each file is being copied into the temp directory:
```
			xmlFileName := filepath.Base(r)
			destFile := filepath.Join(tempRuleDir, xmlFileName)
			err := copyFileContents(r, destFile)
```
In this line mentioned, all of the files passed from input should then be in this dir. Then the dir is copied to volumes.
I tested this scenario, and it looks like passing multiple rules files works as expected.

File: cmd/shimconvert.go (hunk 1 above)
Comment (eemcmullan) [reply cmd/shimconvert.go:55]: Looks like `w.log` requires an error type to return an error, which this condition does not have. Unless I'm missing something?

Comment (pranavgaikwad) [reply cmd/shimconvert.go:55]: you can use Info()

**Review by pranavgaikwad — APPROVED** [review]

**Pull request merged**